hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java

```diff
@@ -743,14 +768,14 @@ public void stop() {
         return;
       }
       balancerRunning = false;
-      currentBalancingThread.interrupt();
-      currentBalancingThread.join(1000);
+      if (Thread.currentThread().getId() != currentBalancingThread.getId()) {
+        currentBalancingThread.interrupt();
+        currentBalancingThread.join();
+      }
```

Comment:
@lokeshj1703 
> Why do we need isBalancerRunning check before calling stop method? It should be ok to call stop directly.

I was thinking of this case:
Suppose `stop()` is called by the CLI. In the stop method, `balancerRunning` would be set to false and `currentBalancingThread` would be interrupted. Then `join()` will wait for `currentBalancingThread` to die while holding the lock. If `currentBalancingThread` now calls `stop()`, it will get disabled since the lock isn't available. This is a possible deadlock.

Reply:
It is a Reentrant lock.

Reply:
In that case, move join call outside the lock.

Reply:
You will need to make sure that start does not succeed until join has completed for the older thread.

Reply:
How about instead of calling lock(), we use tryLock()? If tryLock() returns false then the lock is either held in the start or the stop method. In both the cases, we can just return from the stop method.

Reply:
Another way could be moving the 
`if (!balancerRunning) {
        LOG.info("Container Balancer is not running.");
        return;
      }`
block before the lock is acquired in `stop`.